cmd/pubsub/receive_adapter/main.go

```diff
@@ -72,26 +143,53 @@ func main() {
 		}
 	}
 
-	tracingConfig, err := tracingconfig.JSONToConfig(startable.TracingConfigJson)
+	tracingConfig, err := tracingconfig.JSONToConfig(env.TracingConfigJson)
 	if err != nil {
 		logger.Error("Failed to process tracing options", zap.Error(err))
 	}
 	if err := tracing.SetupStaticPublishing(logger.Sugar(), "", tracingConfig); err != nil {
 		logger.Error("Failed to setup tracing", zap.Error(err), zap.Any("tracingConfig", tracingConfig))
 	}
 
-	if startable.Project == "" {
-		project, err := metadata.ProjectID()
-		if err != nil {
-			logger.Fatal("failed to find project id. ", zap.Error(err))
-		}
-		startable.Project = project
+	projectID, err := utils.ProjectID(env.Project, metadataClient.NewDefaultMetadataClient())
+	if err != nil {
+		logger.Fatal("Failed to retrieve project id", zap.Error(err))
 	}
 
-	logger.Info("Starting Pub/Sub Receive Adapter.", zap.Any("adapter", startable))
-	if err := startable.Start(ctx); err != nil {
-		logger.Fatal("failed to start adapter: ", zap.Error(err))
+	// Convert base64 encoded json map to extensions map.
+	extensions, err := utils.Base64ToMap(env.ExtensionsBase64)
+	if err != nil {
+		logger.Error("Failed to convert base64 extensions to map: %v", zap.Error(err))
 	}
+
+	ctx = context.WithProjectKey(ctx, projectID)
+	ctx = context.WithTopicKey(ctx, env.Topic)
+	ctx = context.WithSubscriptionKey(ctx, env.Subscription)
```

Comment:
Can these be adapter args instead of passed through ctx? It might be worthwhile to create an adapter args struct since InitializeAdapter already has so many arguments.

Reply:
yes, let me do that. It's kind of ugly as is right now. Thanks!

Reply:
agreed to add in AdapterArgs only the ones used by the adapter. In this case topic. The others are needed for initializing other objects as well